### PR TITLE
Add FunCodec to ESPNet

### PR DESCRIPTION
**Pull request by wyh2000**
## What?

This PR aims to add a new codec architecture FunCodec to ESPnet.
FunCodec is based on SoundStream design. Also, it contains some specific design to learn codec in frequency-domain.

In this repo, we add:
1.  Domain transformation and domain inversion (from time-domain to frequency-domain).
2. Seanet2d to support frequency domain codec learning.
3. Support both mag_angle and mag_phase.

## See also

FunCodec paper: https://arxiv.org/pdf/2309.07405
FunCodec official code: https://github.com/modelscope/FunCodec

### CODE REVIEW

**Comment by wyh2000** [comment]
@ftshijt Hi, this PR is ready to be reviewed.

File: espnet2/gan_codec/shared/encoder/seanet.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def get_norm_module(
         if causal:
             raise ValueError("GroupNorm doesn't support causal evaluation.")
         assert isinstance(module, nn.modules.conv._ConvNd)
+        norm_kwargs.pop("num_groups", None)
```
Comment (ftshijt) [inline espnet2/gan_codec/shared/encoder/seanet.py:81]: Not pretty sure, does it have any side-effect?

**Review by ftshijt — COMMENTED** [review]
Many thanks for your great contribution! The code looks great to me. I will merge it after the minor concern is resolved.

File: espnet2/gan_codec/shared/encoder/seanet.py (hunk 1 above)
Comment (wyh2000) [reply espnet2/gan_codec/shared/encoder/seanet.py]: I modified this part. It should be good now.

**Pull request merged**